File: client/src/entities/humanoid.ts
// Procedural articulated humanoid — the blocky Rivals-style body shared by the
// networked opponent and the practice dummy. Limbs hang from shoulder/hip pivot
// groups so a single X-rotation swings each one; the update() drives a
// speed-scaled contralateral walk/run cycle, a subtle idle breath when still,
// and a tucked airborne pose. No skeletons/assets — all code, zero per-frame
// alloc (only scalar rotations are written).

import * as THREE from 'three';
import { TUNING, damp } from '@rivals/shared';

// Proportions (local space; y=0 is the capsule centre, matching the entity
// transforms). Tuned to fill the ~1.8m standing capsule silhouette.
const TORSO = { w: 0.5, h: 0.7, d: 0.28, y: 0.15 };
const HEAD = { s: 0.34, y: 0.68 };
const ARM = { w: 0.16, h: 0.6, d: 0.18, x: 0.37, pivotY: 0.48 };
const LEG = { w: 0.2, h: 0.7, d: 0.22, x: 0.13, pivotY: -0.2 };

// Gait: phase advances by distance travelled so stride frequency tracks speed;
// amplitudes are reached near sprint speed.
const STEP_PER_M = 1.35; // radians of cycle phase per metre travelled
const RUN_REF = TUNING.movement.sprintSpeed; // speed at which amp saturates
const MOVE_EPS = 0.4; // m/s below which we treat the body as standing
const MAX_LEG = 0.85; // peak leg swing (rad)
const MAX_ARM = 0.55; // peak arm swing (rad)
const ARM_REST = 0.06; // slight forward arm rest so they aren't dead straight
const BREATHE_HZ = 1.4; // idle breathing rate

// Make a limb that hangs below a pivot group, so rotating the pivot swings the
// limb from the shoulder/hip. Returns the pivot to parent + animate.
function limb(
  mat: THREE.MeshStandardMaterial,
  w: number,
  h: number,
  d: number,
  x: number,
  pivotY: number,
): THREE.Group {
  const pivot = new THREE.Group();
  pivot.position.set(x, pivotY, 0);
  const mesh = new THREE.Mesh(new THREE.BoxGeometry(w, h, d), mat);
  mesh.position.set(0, -h / 2, 0); // top of the limb sits at the pivot
  pivot.add(mesh);
  return pivot;
}

export class Humanoid {
  readonly object: THREE.Group;

  private readonly torso: THREE.Group;
  private readonly lArm: THREE.Group;
  private readonly rArm: THREE.Group;
  private readonly lLeg: THREE.Group;
  private readonly rLeg: THREE.Group;

  private phase = 0; // gait cycle phase (advances with distance)
  private time = 0; // wall time (idle breathing)
  private amp = 0; // damped move amplitude 0..1
  private air = 0; // damped airborne blend 0..1

  constructor(mat: THREE.MeshStandardMaterial) {
    this.object = new THREE.Group();

    // Torso group also carries the head + arms so the idle breath lifts them
    // together; legs hang off the root so they stay planted.
    this.torso = new THREE.Group();
    const torsoMesh = new THREE.Mesh(
      new THREE.BoxGeometry(TORSO.w, TORSO.h, TORSO.d),
      mat,
    );
    torsoMesh.position.set(0, TORSO.y, 0);
    const headMesh = new THREE.Mesh(
      new THREE.BoxGeometry(HEAD.s, HEAD.s, HEAD.s),
      mat,
    );
    headMesh.position.set(0, HEAD.y, 0);
    this.torso.add(torsoMesh, headMesh);

    this.lArm = limb(mat, ARM.w, ARM.h, ARM.d, -ARM.x, ARM.pivotY);
    this.rArm = limb(mat, ARM.w, ARM.h, ARM.d, ARM.x, ARM.pivotY);
    this.torso.add(this.lArm, this.rArm);

    this.lLeg = limb(mat, LEG.w, LEG.h, LEG.d, -LEG.x, LEG.pivotY);
    this.rLeg = limb(mat, LEG.w, LEG.h, LEG.d, LEG.x, LEG.pivotY);

    this.object.add(this.torso, this.lLeg, this.rLeg);
  }

  /**
   * Advance the animation.
   * @param speed horizontal speed (m/s) — drives stride freq + amplitude
   * @param grounded false blends toward the tucked airborne pose
   */
  update(dt: number, speed: number, grounded: boolean): void {
    this.time += dt;

    const moving = speed > MOVE_EPS;
    if (moving) this.phase += speed * dt * STEP_PER_M;

    // Damp the move amplitude + air blend so transitions don't snap.
    const targetAmp = moving ? Math.min(speed / RUN_REF, 1) : 0;
    this.amp = damp(this.amp, targetAmp, 9, dt);
    this.air = damp(this.air, grounded ? 0 : 1, 10, dt);

    // --- grounded gait ---
    const swing = Math.sin(this.phase);
    const legA = swing * MAX_LEG * this.amp;
    const armA = -swing * MAX_ARM * this.amp; // arms counter the legs

    // --- airborne pose: front leg tucks up, back leg trails, arms raise ---
    const legAir = 0.7; // both legs rotate forward/up a touch in the air
    const armAir = -0.5; // arms come up

    const a = this.air;
    this.lLeg.rotation.x = legA * (1 - a) + legAir * a;
    this.rLeg.rotation.x = -legA * (1 - a) + legAir * 0.4 * a;
    this.lArm.rotation.x = (ARM_REST + armA) * (1 - a) + armAir * a;
    this.rArm.rotation.x = (ARM_REST - armA) * (1 - a) + armAir * a;

    // --- idle breathing (only when essentially still + grounded) ---
    const stillness = (1 - this.amp) * (1 - this.air);
    const breath = Math.sin(this.time * BREATHE_HZ * Math.PI * 2);
    this.torso.position.y = breath * 0.012 * stillness;
    this.torso.rotation.x = breath * 0.02 * stillness;
  }
}

File: godot/src/Client/Render/Humanoid.cs
// Procedural articulated humanoid — the blocky body shared by the networked
// opponent and the practice dummy. Limbs hang from shoulder/hip pivot groups so a
// single X-rotation swings each. Drives a speed-scaled contralateral walk/run
// cycle, a subtle idle breath, and a tucked airborne pose. Ported from
// client/src/entities/humanoid.ts.

using System;
using Godot;
using WebRivals.Shared;

namespace WebRivals.Client.Render;

public sealed class Humanoid
{
    public readonly Node3D Object;

    private readonly Node3D _torso;
    private readonly Node3D _lArm, _rArm, _lLeg, _rLeg;

    private double _phase = 0;
    private double _time = 0;
    private double _amp = 0;
    private double _air = 0;

    private const double StepPerM = 1.35;
    private static readonly double RunRef = Tuning.Movement.SprintSpeed;
    private const double MoveEps = 0.4;
    private const double MaxLeg = 0.85;
    private const double MaxArm = 0.55;
    private const double ArmRest = 0.06;
    private const double BreatheHz = 1.4;

    public Humanoid(Material mat)
    {
        Object = new Node3D();

        _torso = new Node3D();
        _torso.AddChild(BoxMeshNode(mat, 0.5f, 0.7f, 0.28f, new Vector3(0, 0.15f, 0)));   // torso
        _torso.AddChild(BoxMeshNode(mat, 0.34f, 0.34f, 0.34f, new Vector3(0, 0.68f, 0))); // head

        _lArm = Limb(mat, 0.16f, 0.6f, 0.18f, -0.37f, 0.48f);
        _rArm = Limb(mat, 0.16f, 0.6f, 0.18f, 0.37f, 0.48f);
        _torso.AddChild(_lArm);
        _torso.AddChild(_rArm);

        _lLeg = Limb(mat, 0.2f, 0.7f, 0.22f, -0.13f, -0.2f);
        _rLeg = Limb(mat, 0.2f, 0.7f, 0.22f, 0.13f, -0.2f);

        Object.AddChild(_torso);
        Object.AddChild(_lLeg);
        Object.AddChild(_rLeg);
    }

    private static MeshInstance3D BoxMeshNode(Material mat, float w, float h, float d, Vector3 pos)
    {
        var mi = new MeshInstance3D { Mesh = new BoxMesh { Size = new Vector3(w, h, d) }, MaterialOverride = mat, Position = pos };
        mi.CastShadow = GeometryInstance3D.ShadowCastingSetting.On;
        return mi;
    }

    private static Node3D Limb(Material mat, float w, float h, float d, float x, float pivotY)
    {
        var pivot = new Node3D { Position = new Vector3(x, pivotY, 0) };
        pivot.AddChild(BoxMeshNode(mat, w, h, d, new Vector3(0, -h / 2, 0)));
        return pivot;
    }

    public void Update(double dt, double speed, bool grounded)
    {
        _time += dt;

        bool moving = speed > MoveEps;
        if (moving) _phase += speed * dt * StepPerM;

        double targetAmp = moving ? Math.Min(speed / RunRef, 1) : 0;
        _amp = VMath.Damp(_amp, targetAmp, 9, dt);
        _air = VMath.Damp(_air, grounded ? 0 : 1, 10, dt);

        double swing = Math.Sin(_phase);
        double legA = swing * MaxLeg * _amp;
        double armA = -swing * MaxArm * _amp;

        const double legAir = 0.7;
        const double armAir = -0.5;

        double a = _air;
        _lLeg.Rotation = new Vector3((float)(legA * (1 - a) + legAir * a), 0, 0);
        _rLeg.Rotation = new Vector3((float)(-legA * (1 - a) + legAir * 0.4 * a), 0, 0);
        _lArm.Rotation = new Vector3((float)((ArmRest + armA) * (1 - a) + armAir * a), 0, 0);
        _rArm.Rotation = new Vector3((float)((ArmRest - armA) * (1 - a) + armAir * a), 0, 0);

        double stillness = (1 - _amp) * (1 - _air);
        double breath = Math.Sin(_time * BreatheHz * Math.PI * 2);
        _torso.Position = new Vector3(0, (float)(breath * 0.012 * stillness), 0);
        _torso.Rotation = new Vector3((float)(breath * 0.02 * stillness), 0, 0);
    }
}
